[DRAFT] blog: add start of xarray in bio post (#775)
xarray-contrib/xarray.dev: src/posts/xarray-biology/index.md

---
title: 'Accelerating Biological Analysis with Xarray: Why You Should Try Using It Today'
date: '2025-05-04'
authors:
  - name: Ian Hunt-Isaak
    github: ianhi
summary: 'A discussion of how Xarray fits into Biological analysis workflows'
---

If you are a biologist and work with array data (microscopy images, genomic sequences, or anything else you might currently analyze using NumPy). Then you've probably spent hours juggling metadata, battling unclear axes labels, and asking questions like “Why is there a transpose here?” Imagine a tool that will solve those frustrations for you. `Xarray` is that tool.

In this blog post, I will show the areas of biology where it is already in use and discuss its potential for more use cases. I will also explain why it hasn’t already seen wider adoption and the next steps the `Xarray` and Biology communities can take to increase usage of this powerful tool.

## What is Xarray and Why Should You Use it?

> Biological data almost always has rich context and metadata associated with the actual measurements. For example: sample conditions, genetic modifications in a well, timepoints, spatial coordinates. While `NumPy` is a powerful tool, it has limitations when it comes to working with these datasets. Selecting data based on array indices, rather than the physical values, can be confusing. You know you switched the buffer at 32 minutes, but which array index is that? Similarly, keeping track of which dimension is which can be difficult without labels. You have a five-dimensional array, but there are a few transposes in this code from last week, and now you don’t remember which axis is which in the output. Managing a collection of multiple related arrays with slightly different shapes can be tricky. Imagine sending data into a batch job and trying to keep segmentations and raw images together. Or maybe you’ve tried to follow poorly commented analysis from an interesting paper and gotten lost in the details?

Author: I suspect undergrad and above biologists will know what i mean here, but its still worth it come up with a slightly different example and/or be more explicit in what I mean. Ideally if you can understand then every biologist will.

fwiw: 
It is essentially swapping the liquid around the sample to a different liquid. Which will induce biological changes. Buffer has a precise meaning, but is also often used less precisely as I have here.

> `Xarray`’s data structures provide a solution to all these problems. To quote the docs “Xarray introduces labels in the form of dimensions, coordinates, and attributes on top of raw `NumPy`-like multidimensional arrays, which allows for a more intuitive, more concise, and less error-prone developer experience.”

Reviewer: You're diving into highlighting one feature of xarray here (labelled dimensions), but perhaps it would be better to start with  few bullets for different high-level features of interest, e.g.
- labelled dimensions
- lookups via coordinates
- group related arrays together
- powerful analysis methods
- scale to larger-than-memory datasets
- flexible IO

then expand on them?

Reviewer: But also you don't want to to just recreate the docs here, you want to keep it short and to the point, link to the relevant section of docs, and tie it back to bio use cases.

`Xarray` accomplishes this by providing the [`DataArray`](https://docs.xarray.dev/en/latest/user-guide/data-structures.html#dataarray), which keeps track of its metadata and is coordinate aware (e.g., can hold the values of timepoints in seconds for you).

Here is an example of what a `DataArray` of a small stack of microscopy images might look like:

<RawHTML filePath='/posts/xarray-biology/dataarray-repr.html' />

Just by looking at the `repr` you can probably understand a lot about the experiment without any explanation. You no longer need to mentally keep track of transposes, axis labels, and metadata because you can always check the current state!

> Not only does this make it easier for you to develop your analysis, but it also makes your work much more easily understandable and discoverable by others.

Author: I don't think I made this clear, but I'm getting at the use case of someone shares a notebook with you, and you are trying to decipher how it works. I suspect that this will be very compelling, several people have actually brought this up of their own accord to me. I will make this clearer.

### When Can You Use Xarray?

As great as Xarray sounds, it does have limitations. After all, Xarray is an Array library; it’s in the name! So if your data is tabular, Xarray is probably not the right tool for you. To get a better understanding of this, I recommend reading Ryan Abernathy’s excellent post [Tensors vs Tables](https://earthmover.io/blog/tensors-vs-tables?ref=xarray-bio-blog).

That said, many parts of biology generate array data. And scientists in some areas are already benefiting from Xarray.

### Who is Using Xarray

- [`Sgkit`](https://sgkit-dev.github.io/sgkit/latest/getting_started.html#data-structures) “a toolkit for quantitative and population genetics” uses `Xarray` `Dataset`s as a core data structure. This allows them to keep track of the rich metadata associated with sequences and further uses computational patterns to speed up analyses.
- `Xarray` has also been used for neurophysiology data by the Ecephys package. See a previous blog post [xarray for neurophysiology](https://xarray.dev/blog/xarray-for-neurophysiology) to learn more.

> - The `scverse` family of software uses `anndata`, which, like `Xarray` provides labeled dimensions.

Reviewer: So they aren't actually using Xarray?

Author: Felt weird not to mention given the similarities. Also there is this discussion: https://github.com/scverse/anndata/issues/744

Reviewer 2: yeah they don't but would like to. that's why they've been working on pandas extension array stuff. can we add a link pleasE?

- The `bioio` package, capable of reading a zoo of microscope formats, provides a function to get image stacks data as Xarray object.
- The `Xarray.DataTree` provides a natural representation of multiscale images and can interoperate well with the OME-NGFF model. The [multiscale-spatial-image](https://github.com/spatial-image/multiscale-spatial-image?tab=readme-ov-file#multiscale-spatial-image) project use DataTree for precisely this purpose.

## What has limited adoption by Biologists?

> Given the benefits of switching to `Xarray`, why aren’t more biologists using it? Is it secretly not as good as this blog claims?

Reviewer: I think you should put something like this up at the start. The thesis of this post is that
a) Xarray is a powerful tool for bio
b) Some bio people use it
c) A lot more bio people should use it
d) Ideas for how to make that happen

This thesis should be communicated early on and clearly.

As a long-time user of `Xarray` for microscopy data, I had the same question, so I spent February and April 2025 reaching out to scientists and scientific software developers across various fields of biology to answer it. From those discussions, I’ve found that three core reasons have prevented biologists who are already using Python from adopting `Xarray`.

### Awareness and Examples

Many biologists have just never heard of `Xarray`. This is because the geosciences community created Xarray, and there wasn’t enough community overlap for the word to spread to biologists.

A related issue is that the examples and tutorials are geoscience-centric. I admit that when I first read the xarray docs, I was discouraged and thought it wouldn’t work for my microscopy data. Fortunately, I eventually rediscovered Xarray, mentally translated the examples to biology, and found it incredibly useful.

### Technical Barriers

> Once a potential user is convinced of `Xarray`'s value, they may still face technical barriers. Ranging from rough edges to missing features, however, none are insurmountable. An example of a rough edge is that, as of May 2025, you cannot use integers as keys in a `DataTree`. That is a problem, as integers are a natural key to use when tracking single cell lineages. Rough edges like this one haven’t been smoothed over yet because there has not been a user base of biologists using `Xarray`, discovering them, and raising issues to get them fixed.

Reviewer: This should be two paragraphs. One introducing the idea that technical limitations might be preventing use by bio people, then start enumerating examples in a separate paragraph / list.

> Once a potential user is convinced of `Xarray`'s value, they may still face technical barriers. Ranging from rough edges to missing features, however, none are insurmountable. An example of a rough edge is that, as of May 2025, you cannot use integers as keys in a `DataTree`. That is a problem, as integers are a natural key to use when tracking single cell lineages. Rough edges like this one haven’t been smoothed over yet because there has not been a user base of biologists using `Xarray`, discovering them, and raising issues to get them fixed.

Author: Good question:

Thinking out loud this breaks down into two cases:

1. no user base in a subfield (e.g. crystallography)
2. small user base and usage hasn't proliferated yet: e.g., imaging, sgkit

Until very recently, a missing feature was allowing `Xarray` coordinates to be more flexible than a fully instantiated array. Allowing for analytical transforms and having more flexible coordinates is critical to supporting Volumetric imaging applications. Excitingly, in February 2025, initial support for this was [merged](https://github.com/pydata/xarray/pull/9543)!

Another area of potential improvement is support for sparse arrays. Many biology datasets are quite sparse, so robust support is a key feature. Xarray does have support for sparse arrays, but there are still open issues, for example [Issue 3212](https://github.com/pydata/xarray/issues/3213).

### Data Loading/Lack of Integration

> Finally, we have been limited by a lack of integration with existing software tools. First, in loading the outputs of other tools into `Xarray` (do they have a `to_xarray` method?), and second, in other tools accepting `Xarray` arrays and using the extra features. For example, [Napari](https://napari.org/stable/) has had a long-standing [open issue](https://github.com/napari/napari/issues/14) about using `Xarray` to add extra information to dimension sliders.

Reviewer: Why is there a lack of integration though? Lack of funding? Lack of understanding of the benefits?

## Call To Action

`Xarray` has fantastic potential to accelerate our ability to perform and share biological analyses and accelerate discovery. To realize this potential, there are a few things we as a community need to do.

The first thing is to consider what we mean by “community”. There are two separate communities with some overlap: `Xarray` developers and biologists interested in using `Xarray`.

### What can Xarray Do

As a community of contributors to `Xarray`, we need to:

- **Support**: Be responsive to technical limitations and bugs in Xarray that affect biologists
- **Educate**: cross-post about improvements in Xarray to spaces with biologists

### What can Biologists Do?

**Try** using `Xarray` for your data. Find the smallest, most straightforward way to get your data into Xarray and start using it.

Make sure to **ask for help** when you need it. To do so:

- Post on forums (e.g., [image.sc](https://forum.image.sc/), or the [Xarray Github Discussions](https://github.com/pydata/xarray/discussions))
- Join the new Xarray for Biology office hours [TODO LINK]

> - Book one on one time with Ian [TODO LINK]

Author: Need to create a link for these. Not quite sure best practice for this. github discussion?

Reviewer 2: Use calendly for now IMO.

Share how you did it with your colleagues and report bugs and feature requests to Xarray.

If you have a success story, then you should share it! For example, via a short blog post here.

### What can the Xarray-Bio Community do?

Anyone already using `Xarray` to work with biology data is uniquely well-positioned to help speed adoption. We have three core tasks to grow the number of us.

**Build** and then **share** tools that use `Xarray` to do useful things, such as building domain-specific visualization tools that use Xarray metadata to build the visualizations. Or solving any problems that you run into, if it bothers you, it likely bothers someone else and they’ll be happy to use your solution.

**Contribute** by commenting on Xarray issues with how they affect biology use cases, and working on bug fixes and new features.

**Support** other biologists learning to use Xarray. Respond to forum posts and help budding users, write and share small examples of using Xarray with biology data. Teach tutorials to your peers.

> My current role is an “Xarray community Developer” focusing on biological applications. So for my part, I’m always happy to talk to you about whether Xarray might be a good fit for your biology data. Please reach out if you have a question! I’m `@ianhi` on most platforms. You can also join our new Xarray in Biology office hours [LINK], or book some time with me to talk Xarray and Biology [here](https://calendly.com/ian-earthmover/30min).

Reviewer: Your role is an important part of the narrative arc of the post.

Reviewer 3: Agree with this. Let's move your own introduction to the intro. The arc could be:

- Hi, I'm Ian. I'm the new Xarray community dev, I'll be focusing on bio applications. btw, thanks CZI for the funding.
- In this post, I'll be surveying the landscape of Xarray for bio applications. This post is for you if: ...

- Ability to scale to larger than memory datasets via [integration with Dask](https://docs.xarray.dev/en/stable/user-guide/dask.html)
- Flexible [I/O framework](https://docs.xarray.dev/en/stable/user-guide/io.html) with potential for better integration with a zoo of bio formats

### When Can You Use Xarray?

As great as Xarray sounds, it does have limitations. Xarray is an array library; it's in the name! So, if your data is tabular, Xarray is probably not the right tool for you. To better understand this, I recommend reading Ryan Abernathy's excellent post [Tensors vs. Tables](https://earthmover.io/blog/tensors-vs-tables?ref=xarray-bio-blog).

That said, many parts of biology generate array data. And scientists in some areas are already benefiting from Xarray.

### Who is Using Xarray

- [`Sgkit`](https://sgkit-dev.github.io/sgkit/latest/getting_started.html#data-structures) "a toolkit for quantitative and population genetics" uses `Xarray` `Dataset`'s as a core data structure. `Xarray` allows them to keep track of the rich metadata associated with sequences and further uses computational patterns to speed up analyses.

>   - `Xarray`'s interoperability with Zarr makes it a promising way to interact with [`VCF Zarr`](https://www.biorxiv.org/content/10.1101/2024.06.11.598241v3)

Reviewer 2: can we expand `VCF`?

> - `Xarray` has also been used for neurophysiology data by the Ecephys package. See a previous blog post [xarray for neurophysiology](https://xarray.dev/blog/xarray-for-neurophysiology) to learn more.

Reviewer 2: link to Ecephys?

### Technical Barriers

Once a potential user is convinced of `Xarray`'s value, they may still face technical barriers. Ranging from rough edges to missing features, however, none are insurmountable. These barriers can exist in `Xarray` or in other tools in their workflow not accepting `Xarray` as input.

An example of a rough edge in `Xarray` is that, as of May 2025, you cannot use integers as keys in a `DataTree`. That is a problem, as integers are a natural key when tracking single-cell lineages. Rough edges like this one haven't been smoothed over yet because there has not been a large user base of biologists using `Xarray` and raising issues when they encounter problems.

Until recently, a missing feature was the ability to have `Xarray` coordinates that are more flexible than a fully instantiated array. Allowing for analytical transforms and having more flexible coordinates is critical to supporting volumetric imaging applications. Excitingly, in February 2025, initial support for this was [merged](https://github.com/pydata/xarray/pull/9543)!

> Another area of potential improvement is support for sparse arrays. Many biology datasets are sparse, so robust support is a key feature. Xarray does support sparse arrays, but there are still open issues, for example, [Issue 3212](https://github.com/pydata/xarray/issues/3213).

Reviewer 2: is this a real blocker? that's a random tracking issue. Here, we should ask readers to open an issue for sparse-related features

The first thing is to consider what we mean by "community." There are two separate communities with some overlap: `Xarray` developers and biologists interested in using `Xarray`.

### What can Xarray Do

As a community of contributors to `Xarray`, we need to:

- **Support**: Be responsive to technical limitations and bugs in Xarray that affect biologists
- **Educate**: cross-post about improvements in Xarray to spaces with biologists

### What can Biologists Do?

**Try** using `Xarray` for your data. Find the smallest, most straightforward way to get your data into Xarray and start using it.

Make sure to **ask for help** when you need it. To do so:

- Post on forums (e.g., [image.sc](https://forum.image.sc/), or the [Xarray Github Discussions](https://github.com/pydata/xarray/discussions))
- Join the new Xarray for Biology office hours [TODO LINK]
- Book one on one time with Ian [TODO LINK]

**Share** how you did it with your colleagues and report bugs and feature requests to Xarray.

> If you have a success story, then you should tell people about it! For example, by submitting a short blog post here.

Reviewer 2: let's link to the issues tab on this website

### What can the Xarray-Bio Community do?

Anyone already using `Xarray` to work with biology data is uniquely well-positioned to help speed adoption. We have three core tasks to grow the number of us.

**Build** and then **share** tools that use `Xarray` to do useful things, such as building domain-specific visualization tools that use Xarray metadata to build the visualizations. Or solving any problems that you run into; if it bothers you, it likely bothers someone else, and they'll be happy to use your solution.

**Contribute** by commenting on Xarray issues, explaining how they affect biology use cases, and working on bug fixes and new features.

**Support** other biologists learning to use Xarray. Respond to forum posts and help budding users, write and share small examples of using Xarray with biology data. Teach tutorials to your peers.

> As part of my role as an "Xarray community Developer" I'm always happy to talk to you about whether Xarray might be a good fit for your biology data. Please reach out if you have a question! I'm `@ianhi` on most platforms. You can also join our new Xarray in Biology office hours [LINK], or book some time with me to talk Xarray and Biology [here](https://calendly.com/ian-earthmover/30min).

Reviewer 2: Let's do a Xarray-in-bio sprint at SciPy? And call it out here? For example, we could create a simple google form and ask people to sign up. we could also advertise that form in a banner on the xarray.dev website

Author: Great call. There isn't any limitation on sprints at scipy right, they are just self organized?

Reviewer 2: Yes IIRC

Reviewer: Yes you don't even have to decide until the morning of, when you stand up and advertise your sprint

## Looking Forward

`Xarray` has already demonstrated its value to biologists when it is used. If we continue to work toward using this tool, it will be a significant step toward having standardized, interoperable, metadata-rich datasets, maintainable workflows, and easier interactive analysis. This will benefit the geosciences community with contributions and benefit biologists by avoiding duplication of effort.

> This blog post addressed biology as a broad topic. In future posts, I will explore domain-specific insights and examples in fields such as microscopy, cryo-EM/ET, x-ray crystallography data, high-throughput screening, etc... If you are interested in talking about this, helping write posts, or otherwise getting involved, please don't hesitate to reach out.

Reviewer 2: is there a way to find your email?

---
title: 'Accelerating Biological Analysis with Xarray: Why You Should Try Using It Today'
date: '2025-05-04'
authors:
  - name: Ian Hunt-Isaak
    github: ianhi
summary: 'A discussion of how Xarray fits into Biological analysis workflows'
---

I'm Ian, a multimodal microscopist, and the new "Xarray Community Developer." In this role, funded by the Chan Zuckerberg Institute, I work full-time to support the usage of Xarray in biological and biomedical applications.

> I ended up here because, five years in graduate school, I was sitting at my computer frustratedly trying to keep track of microscopy image metadata as I passed images through complicated, larger-than-memory jobs on the cluster. Which axis was what? Where did the exposure times go? I know I changed the media at 32 minutes, but what time index was that? Desperate for a solution, I found `Xarray`, seemingly a solution to all my woes. But I left thinking it was not for me as the examples were geoscience-centric. Fortunately, I eventually rediscovered `Xarray`, mentally translated the examples to biology, and found it incredibly useful. It is so useful that it since served as my default tool for all analyses in both structural and microbiology.

Reviewer 2: Let's just use Xarray, NumPy etc. and not stick the project names in `code` formatting

My story is not unique. You have likely experienced similar challenges when working with array data (e.g., microscopy images, genomic sequences, or anything else you might currently analyze using `NumPy`). This post is for you whether you have heard of `Xarray` before or not.

> In this post, I will introduce the concepts of `Xarray` at a high level with biological context and give examples where it is already in use. Then, based on interviews I conducted with scientists and scientific software developers from across biology I will explain what has limited adoption. Finally, I will describe what we (Biologists and Xarray contributors) can do to increase adoption.

Reviewer 2: I suggest rephrasing this to start with a short description of what you've down over the past two months, and present this post as a summary.

```suggestion
I have spent the last XXXXX.... 

In this post, I will introduce Xarray at a high level with biological context and give examples where it is already in use. Then, based on interviews I conducted with scientists and scientific software developers from a number of biology subfields I will explain what has limited adoption. Finally, I will describe what we (biologists and Xarray contributors) can do to increase adoption.
```

When you collect data, you likely collect more than just one type. You may have sequence data in addition to growth curves, fluorescence data, or some kind of spectral data. If you collect multiple arrays of data that share some dimensions (e.g., time) but not all, how do you keep them organized? They can't fit in one array as they don't share all their dimensions!

Happily, `Xarray` solves this problem for us as well! The [`Dataset`](https://docs.xarray.dev/en/latest/user-guide/data-structures.html#dataset) object stores collections of related `DataArray`'s. For example, for our microscopy image stack, we might also create a segmentation layer that will share some of the dimensions of our image (`time`, `X`, `Y`, [`Z`]) but not all (`channel`, [`Z`]) and it would be great to keep them together.

<RawHTML filePath='/posts/xarray-biology/dataset-repr.html' />

> Finally, the [`DataTree`](https://docs.xarray.dev/en/latest/user-guide/data-structures.html#dataset) object allows you to keep trees of related datasets, for example, a representation of multiscale images allows for natural support of [OME-NGFF](https://www.nature.com/articles/s41592-021-01326-w) data.

Reviewer 2: ```suggestion

#### DataTree
Finally, the [`DataTree`](https://docs.xarray.dev/en/latest/user-guide/data-structures.html#dataset) object allows you to keep trees of related datasets, for example, a representation of multiscale images allows for natural support of [OME-NGFF](https://www.nature.com/articles/s41592-021-01326-w) data.
```

<RawHTML filePath='/posts/xarray-biology/datatree-repr.html' />

Having labeled dimensions and the ability to organize multiple arrays is often reason enough to start using Xarray. But there are more benefits:

- Readable and easy to use [indexing and selecting semantics](https://docs.xarray.dev/en/stable/user-guide/indexing.html)
  - e.g., Max projected GFP from minutes 15 and 30: `array.sel(channel='GFP', time=(15,30)).max(‘Z’)`
- [Data merging](https://docs.xarray.dev/en/stable/user-guide/combining.html) based on dimensions and coordinates
- Built-in [vizualization tooling](https://docs.xarray.dev/en/stable/user-guide/plotting.html)
- Powerful [computational Patterns](https://tutorial.xarray.dev/intermediate/01-high-level-computation-patterns.html) such as `groupby`, `coarsen`, `resample`
- Ability to scale to larger than memory datasets via [integration with Dask](https://docs.xarray.dev/en/stable/user-guide/dask.html)
- Flexible [I/O framework](https://docs.xarray.dev/en/stable/user-guide/io.html) with potential for better integration with a zoo of bio formats
- Arbitrary array types including `pydata/sparse` arrays.

### When Can You Use Xarray?

> As great as Xarray sounds, it does have limitations. Xarray is an array library; it's in the name! So, if your data is tabular and the tabular ecosystem is working well for you then keep using that!

Reviewer 2: Xarray does handle collections of 1D arrays perfectly well though.

## Call To Action

`Xarray` has fantastic potential to accelerate our ability to perform and share biological analyses and accelerate discovery. To realize this potential, there are a few things we as a community need to do.

The first thing is to consider what we mean by "community." There are two separate communities with some overlap: `Xarray` developers and biologists interested in using `Xarray`.

### What can Xarray Do

As a community of contributors to `Xarray`, we need to:

- **Support**: Be responsive to technical limitations and bugs in Xarray that affect biologists
- **Educate**: cross-post about improvements in Xarray to spaces with biologists

### What can Biologists Do?

**Try** using `Xarray` for your data. Find the smallest, most straightforward way to get your data into Xarray and start using it.

> Make sure to **ask for help** when you need it. To do so:

Reviewer 2: ```suggestion
**Ask for help** when you need it. To do so:
```

matches the rest...

> 

Reviewer 2: ```suggestion
```

- Join the new Xarray for Biology [office hours](https://xarray-contrib.github.io/xarray-for-bio/getting-help.html#office-hours)
- Book one on one time with Ian [here](https://calendar.app.google/xdjhynWycxE3okk68).

**Share** how you did it with your colleagues and report bugs and feature requests to Xarray.

If you have a success story, then you should tell people about it! For example, by submitting a short blog post [here](https://github.com/xarray-contrib/xarray.dev/issues).

### What can the Xarray-Bio Community do?

Anyone already using `Xarray` to work with biology data is uniquely well-positioned to help speed adoption. We have three core tasks to grow the number of us.

**Build** and then **share** tools that use `Xarray` to do useful things, such as building domain-specific visualization tools that use Xarray metadata to build the visualizations. Or solving any problems that you run into; if it bothers you, it likely bothers someone else, and they'll be happy to use your solution.

**Contribute** by commenting on Xarray issues, explaining how they affect biology use cases, and working on bug fixes and new features.

> - **SciPy 2025 Sprint!** If you will be attending the [SciPy 2025](https://www.scipy2025.scipy.org/) conference join our Xarray for Biology Sprint! It's not required but if you are interested let us knnow via this [form](https://docs.google.com/forms/d/e/1FAIpQLSeGvTLONF-24V7z2HoACm4MhEr82c2V-VIzA9eqM9-jt-Xh8g/viewform?usp=sharing&ouid=111570313164368772519)

Author: Are there admonitions or similar on this blog? Would be nice to put it in a card to mark it as separate

---
title: 'Accelerating Biological Analysis with Xarray'
date: '2025-06-04'
authors:
  - name: Ian Hunt-Isaak
    github: ianhi
summary: 'A discussion of how Xarray fits into Biological analysis workflows'
---

**SciPy 2025 Sprint!** If you will be attending the [SciPy 2025](https://www.scipy2025.scipy.org/) conference join our Xarray for Biology Sprint! It's not required but if you are interested let us knnow via this [form](https://docs.google.com/forms/d/e/1FAIpQLSeGvTLONF-24V7z2HoACm4MhEr82c2V-VIzA9eqM9-jt-Xh8g/viewform?usp=sharing&ouid=111570313164368772519)

> 

Reviewer 3: ```suggestion

----------
```

I'm Ian, a multimodal microscopist, and the new "Xarray Community Developer." In this role, funded by the Chan Zuckerberg Institute, I work full-time to support the usage of Xarray in biological and biomedical applications.

I ended up here because, five years in graduate school, I was sitting at my computer frustratedly trying to keep track of microscopy image metadata as I passed images through complicated, larger-than-memory jobs on the cluster. Which axis was what? Where did the exposure times go? I know I changed the media at 32 minutes, but what time index was that? Desperate for a solution, I found Xarray, seemingly a solution to all my woes. But I left thinking it was not for me as the examples were geoscience-centric. Fortunately, I eventually rediscovered Xarray, mentally translated the examples to biology, and found it incredibly useful. It is so useful that it since served as my default tool for all analyses in both structural and microbiology.

> My story is not unique. You have likely experienced similar challenges when working with array data (e.g., microscopy images, genomic sequences, or anything else you might currently analyze using NumPy). This post is for you whether you have heard of Xarray before or not.

Author: ADjusted

I have spent the last several months interviewing scientists and software developers across many fields of Biology. I also contributed biology related fixes to Xarray and Zarr, attended conferences and studied existing use cases of Xarray. These conversations and experiences are the basis of the research I have been doing on the how Xarray can be used in biological applications.

This post contains a summary of my findings. I will introduce the concepts of Xarray at a high level with biological context and give examples where it is already in use. Then, based on the interviews I conducted I will explain what has limited adoption. Finally, I will describe what we (Biologists and Xarray contributors) can do to increase adoption.

## What is Xarray and Why Should You Use it?

Biological data almost always has rich context and metadata associated with the actual measurements. For example, sample conditions, genetic modifications, buffer composition per well in a plate, time points, and spatial coordinates. While NumPy is a powerful tool, it has limitations when it comes to working with these datasets.

> Selecting data based on array indices rather than the physical values can be confusing. You know you switched the growth medium at 32 minutes, but which array index is that? Similarly, keeping track of which dimension is which can be difficult without labels. You have a five-dimensional array, but there are a few transposes in this code from last week, and now you don't remember which axis is which in the output. Managing a collection of multiple related arrays with slightly different shapes can be tricky. Imagine sending data into a batch job and trying to keep segmentations and raw images together. Or maybe you've tried to follow poorly commented analysis from an interesting paper and gotten lost in the details?

Reviewer 4: I find this convincing, but I wonder whether it can be shortened a bit.

- The `Xarray.DataTree` provides a natural representation of multiscale images and can interoperate well with the OME-NGFF model. The [multiscale-spatial-image](https://github.com/spatial-image/multiscale-spatial-image?tab=readme-ov-file#multiscale-spatial-image) project uses `DataTree` for precisely this purpose.

## What has limited adoption by Biologists?

Given the benefits of switching to `Xarray`, why aren't more biologists using it? Is it secretly not as good as this blog claims? I spent February and April 2025 reaching out to scientists and scientific software developers across various fields of biology to find out. From those discussions, I've found three core reasons have prevented biologists who are already using Python from adopting `Xarray`.

### Awareness and Examples

Many biologists have never heard of `Xarray`. This is because the geosciences community created Xarray, and there wasn't enough community overlap for the word to spread to biologists.

A related issue is that the examples and tutorials are geoscience-centric, which was the source of my discouragement when I first read the Xarray docs.

### Technical Barriers

> Once a potential user is convinced of `Xarray`'s value, they may still face technical barriers. Ranging from rough edges to missing features, however, none are insurmountable. These barriers can exist in `Xarray` or in other tools in their workflow not accepting `Xarray` as input.

Reviewer 4: ```
Even though they'd like to use `Xarray`, potential users may face technical barriers in the form of rough edges or missing features. These barriers may be in `Xarray` or the other tools in their workflow.
```

> An example of a rough edge in `Xarray` is that, as of May 2025, you cannot use integers as keys in a `DataTree`. That is a problem, as integers are a natural key when tracking single-cell lineages. Rough edges like this one haven't been smoothed over yet because there has not been a large user base of biologists using `Xarray` and raising issues when they encounter problems.

Reviewer 4: ```
An example of a rough edge in `Xarray` is that you cannot use integers as keys in a `DataTree`. That is a problem, as integers are a natural key when tracking single-cell lineages. Rough edges like this one haven't been smoothed over yet because there has not been a large user base of biologists raising issues when they encounter problems.
```

> Until recently, a missing feature was the ability to have `Xarray` coordinates that are more flexible than a fully instantiated array. Allowing for analytical transforms and having more flexible coordinates is critical to supporting volumetric imaging applications. Excitingly, in February 2025, initial support for this was [merged](https://github.com/pydata/xarray/pull/9543)!

Reviewer 4: As an outsider, I don't know what `coordinates that are more flexible than a fully instantiated array` means.

> Another area of potential improvement is support for sparse arrays. Many biology datasets are sparse, so robust support is a key feature. Xarray does currently support sparse arrays so if you have tried using them and ran into problems please [open an issue](https://github.com/pydata/xarray/issues/new/choose).

Reviewer 4: From my perspective, this is a very important feature. Can you cite an important example data set including its memory footprint when represented by sparse and dense arrays? Perhaps something from the DNA / RNA sequencing domain?

### Lack of Integration

Finally, we have been limited by a lack of integration with existing software tools. This is partially a consequence of prioritization from downstream developers. It's a classic chicken and egg problem: Users with `Xarray` data aren't asking for integration, so why spend time incorporating it? Similarly, the software doesn't support `Xarray`, so why spend time putting your data in `Xarray`?

> This comes up in two ways. First, whether tools can output their internal format to `Xarray`, and second, whether tools accept `Xarray` arrays and make use of the extra features. For example, [Napari](https://napari.org/stable/) has had a long-standing [open issue](https://github.com/napari/napari/issues/14) about using `Xarray` to add extra information to dimension sliders.

Reviewer 4: What's Napari?

## Call To Action

Xarray has fantastic potential to accelerate our ability to perform and share biological analyses and accelerate discovery. To realize this potential, there are a few things we as a community need to do.

The first thing is to consider what we mean by "community." There are two separate communities with some overlap: Xarray developers and biologists interested in using Xarray.

### What can Xarray Do

As a community of contributors to Xarray, we need to:

- **Support**: Be responsive to technical limitations and bugs in Xarray that affect biologists
- **Educate**: cross-post about improvements in Xarray to spaces with biologists

### What can Biologists Do?

**Try** using Xarray for your data. Find the smallest, most straightforward way to get your data into Xarray and start using it.

Make sure to **ask for help** when you need it. To do so:

- Post on forums (e.g., [image.sc](https://forum.image.sc/), or the [Xarray Github Discussions](https://github.com/pydata/xarray/discussions))

>   As part of my role as an "Xarray community Developer" I'm always happy to talk to you about whether Xarray might be a good fit for your biology data. Please reach out if you have a question! I'm `@ianhi` on most platforms. You can also join our new Xarray in Biology

Author: ```suggestion
  As part of my role as an "Xarray community Developer" I'm always happy to talk to you about whether Xarray might be a good fit for your biology data. Please reach out if you have a question! I'm `@ianhi` on most platforms.
```
